sycl/include/sycl/detail/spirv.hpp

```diff
@@ -775,217 +803,315 @@ inline uint32_t membermask() {
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleDown(g, x[s], delta);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    id<1> TargetLocalId = g.get_local_id();
+    // ID outside the group range is UB, so we just keep the current item ID
+    // unchanged.
+    if (TargetLocalId[0] + delta < g.get_local_linear_range())
+      TargetLocalId[0] += delta;
+    uint32_t TargetId = MapShuffleID(g, TargetLocalId);
+    return __spirv_GroupNonUniformShuffle(group_scope<GroupT>::value,
+                                          convertToOpenCLType(x), TargetId);
+  } else {
+    // Subgroup.
+    return __spirv_SubgroupShuffleDownINTEL(convertToOpenCLType(x),
+                                            convertToOpenCLType(x), delta);
+  }
 #else
-  return __nvvm_shfl_sync_down_i32(membermask(), x, delta, 0x1f);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT>) {
+    return __nvvm_shfl_sync_down_i32(detail::ExtractMask(detail::GetMask(g))[0],
+                                     x, delta, 0x1f);
+  } else {
+    return __nvvm_shfl_sync_down_i32(membermask(), x, delta, 0x1f);
+  }
 #endif
 }
 
-template <typename T>
-EnableIfNativeShuffle<T> SubgroupShuffleUp(T x, uint32_t delta) {
+template <typename GroupT, typename T>
+EnableIfNativeShuffle<T> ShuffleUp(GroupT g, T x, uint32_t delta) {
 #ifndef __NVPTX__
-  return __spirv_SubgroupShuffleUpINTEL(convertToOpenCLType(x),
-                                        convertToOpenCLType(x), delta);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT> &&
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
+    // TODO: Remove when IGC bug is fixed.
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleUp(g, x[s], delta);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    id<1> TargetLocalId = g.get_local_id();
+    // Underflow is UB, so we just keep the current item ID unchanged.
+    if (TargetLocalId[0] >= delta)
+      TargetLocalId[0] -= delta;
+    uint32_t TargetId = MapShuffleID(g, TargetLocalId);
+    return __spirv_GroupNonUniformShuffle(group_scope<GroupT>::value,
+                                          convertToOpenCLType(x), TargetId);
+  } else {
+    // Subgroup.
+    return __spirv_SubgroupShuffleUpINTEL(convertToOpenCLType(x),
+                                          convertToOpenCLType(x), delta);
+  }
 #else
-  return __nvvm_shfl_sync_up_i32(membermask(), x, delta, 0);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT>) {
+    return __nvvm_shfl_sync_up_i32(detail::ExtractMask(detail::GetMask(g))[0],
+                                   x, delta, 0);
```

Comment:
@JackAKirk - Can you think of a reason why this would fail for the new test cases for ballot_group only?

Reply:
I think it is failing because the delta does not relate to the true delta only considering set bits in the mask which is what is desired. I think could also fail for opportunistic_group, but guess this depends on the test/execution.

I think for these cases if you replace
```
    return __nvvm_shfl_sync_up_i32(detail::ExtractMask(detail::GetMask(g))[0],
                                   x, delta, 0);
``` 
with a call to
```
non_uniform_shfl
```
defined here:

https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/experimental/cuda/non_uniform_algorithms.hpp#L151

as e.g. 
```
unsigned localSetBit = g.get_local_id()[0] + 1;
int unfoldedSrcSetBit = localSetBit + delta;
auto MemberMask = detail::ExtractMask(detail::GetMask(g))[0];
return non_uniform_shfl(g, MemberMask, x,
                                __nvvm_fns(MemberMask, 0, unfoldedSrcSetBit));
```

Then it should work.

Then it should work.

Reply:
Actually thinking about it more, in this case I think the sign of delta might have to be changed for ballot_group/opportunistic group, when using that `non_uniform_shfl` function.
e.g. in my above message you might have to replace
```
int unfoldedSrcSetBit = localSetBit + delta;
```

with:

```
int unfoldedSrcSetBit = localSetBit - delta;
```

In any case I think one of those should work. I can't remember the docs offhand for which version is correct (basically whether the semantic is send to idx or receive from). I think it is the `- delta` version!

Reply:
Thanks a ton, Jack! Let's give it a try! 😄

Reply:
Sadly it doesn't seem that it changed anything, but it also doesn't seem like we cann `non_uniform_shfl` here. I wonder if it is immediately usable or if it has been specialized for the use-case with reduce and scan. Either way, I do not have an easy way of testing fixes, so would you be okay with me disabling the subset of checks for CUDA and opening an issue so you can have a look when possible? (Tag @npmiller )

Reply:
Sure, no worries. I think I know how to fix it.